modules/javafx.graphics/src/main/java/com/sun/javafx/application/preferences/ChangedValue.java

```diff
@@ -0,0 +1,73 @@
+/*
+ * Copyright (c) 2023, Oracle and/or its affiliates. All rights reserved.
+ * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
+ *
+ * This code is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 2 only, as
+ * published by the Free Software Foundation.  Oracle designates this
+ * particular file as subject to the "Classpath" exception as provided
+ * by Oracle in the LICENSE file that accompanied this code.
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+
+package com.sun.javafx.application.preferences;
+
+import java.util.Arrays;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Objects;
+
+/**
+ * Contains information about a changed value.
+ */
+public record ChangedValue(Object oldValue, Object newValue) {
+
+    /**
+     * Returns a map that contains the new or changed mappings of {@code current} compared to {@code old}.
+     * A value has changed if {@link Objects#equals(Object, Object)} or {@link Arrays#equals(Object[], Object[])}
+     * returns {@code false} when invoked with the old and new value.
+     *
+     * @param old the old mappings
+     * @param current the current mappings
+     * @return a mapping of keys to changed values
+     */
+    public static Map<String, ChangedValue> getEffectiveChanges(Map<String, Object> old, Map<String, Object> current) {
```

Comment:
this class does not handle addition of keys in `current` - should we explain this fact in this method and/or the class javadoc?

Reply:
It doesn't handle _removals_ of keys in `current`, which is explained in the method documentation:

"Returns a map that contains the _new or changed_ mappings of _current_ compared to _old_.
A value has changed if _Objects#equals_ or _Arrays#equals_ returns _false_ when invoked with the old and new value."

If you think this is not clear enough, I can try to improve the wording.

Reply:
is it actually possible to have keys removed at runtime?

Reply:
The set of preferences that is reported by a platform is hard-coded in the native platform implementation, and depends on the operating system version. It might indeed be the case that an OS upgrade/downgrade could result in a different set of reported properties, but can this happen while a JavaFX applications remains running?

In any case, even if that were to happen, the current behavior is that a preference that was once reported will not be removed.

Reply:
Looks like I spoke too soon. Mappings can actually be removed at runtime, one such example is `Windows.SPI.HighContrastColorScheme`, which is only available if `Windows.SPI.HighContrast` is `true`.

This doesn't change the `ChangedValue` implementation, and there are tests that cover the key removal scenario.

Reply:
if that's the case, isn't it possible to have the same key added at runtime?  this scenario cannot be handled by the current code, as far as I can tell.

Reply:
I'm not sure if I understand what you mean that the same key can be added at runtime. The protocol is as follows:

1. The native toolkit reports either _all_ preferences, or the _changed_ preferences (or a mixture of both). If a report includes only changed preferences, this _doesn't_ imply that unreported preferences (that were reported earlier) are removed.
2. `PlatformPreferences.update` computes the difference between the _currently known_ preferences, and the _reported_ preferences. This is what `ChangedValue.getEffectiveChanges` does. A mapping that is not included in the reported preferences doesn't mean that it was removed; if the native toolkit wants to signal removal, it needs to map the respective key to `null` instead.
3. If there are any _effective changes_ (i.e. changes where the current value is actually different from the last known value), change notifications are fired. This includes additions (where new keys are seen for the first time), as well as removals (where a known key is now mapped to `null`).

This protocol gives native toolkits the option to choose how preferences are reported, some toolkits may always want to report all preferences, while other toolkits may only want to report the changed preferences. The burden of keeping track of the total set of preferences is placed on `PlatformPreferences`, not on the native toolkit. This allows us to implement the book-keeping logic in a single place, rather than in each toolkit individually.

Reply:
thanks for clarification. I misunderstood your earlier comment.